Fix problems with leading '\' in FQCNs

VolCh opened the pull request:
I found a inconsistent behavior about classname in native \Reflection* and \Go\ParserReflection\Reflection* 

Is it unexpected behavior? The native Reflection* constructors drop the leading slash in the string class name parameter. I made changes in constructors to consist behavior and pass thees tests.  It's ready to push. Also it closes #80.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/9946245a-5eb1-49ef-a4de-8d7ee8a065d6) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/9946245a-5eb1-49ef-a4de-8d7ee8a065d6) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/0e575ad4-a3ca-46bf-84e3-7c4cb876dbba) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/dd45fc64-747f-4db1-b6cf-0e0f2ce06db4) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/0e575ad4-a3ca-46bf-84e3-7c4cb876dbba) completed: **No new issues**

aik099 commented on tests/ReflectionClassTest.php:
```diff
@@ -290,4 +290,37 @@ protected function getGettersToCheck()
+    /**
+     * Tests getName() method
+     *
+     * @dataProvider getClassNamesToCheck
```
If I'm not mistaken, then `@dataProvider` annotation should go after everything else. Or do as done in other places.

VolCh replied:
Technically it doesn't matter where.  II use style of ReflectionClassTest::testGetModifiers() & ReflectionClassTest::testReflectionMethodParity()

aik099 commented on tests/ReflectionClassTest.php:
```diff
@@ -290,4 +290,37 @@ protected function getGettersToCheck()
+     *
+     * @param string $className
+     */
+    public function testClassName($className) {
```
The `{` should be on next line as in other places.

aik099 commented on tests/ReflectionClassTest.php:
```diff
@@ -290,4 +290,37 @@ protected function getGettersToCheck()
+     */
+    public function testClassName($className) {
+
+        $this->assertSame(
```
Interesting. I the `getName` method result is compared in `testReflectionMethodParity` test, but one of 2 things happens:

1. either `->getName()` result is correct AND `->name` isn't correct
2. or class names list used for testing isn't complete

Likely you should add classes from global namespace to these tests as well.

loren-osborn replied:
Should we add FQ names for classes and functions as distinct cases in the `@dataProvider`s for `ReflectionClassTest` and `ReflectionFunctionTest`. (We should also similarly add case-insensitivity test cases to the `@dataProvider`s for `ReflectionFunctionTest` and `ReflectionMethodTest` While we're at it.)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/35326305-5648-4564-8d38-5c60bd45faba) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/35326305-5648-4564-8d38-5c60bd45faba) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/8d74194a-4ba0-4747-9300-9fb2c689067a) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/8d74194a-4ba0-4747-9300-9fb2c689067a) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/909a4250-74e4-481a-8b84-74b02fff74fc) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/9da289d3-6b86-40e2-9137-34456d836adb) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/909a4250-74e4-481a-8b84-74b02fff74fc) completed: **No new issues**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/9da289d3-6b86-40e2-9137-34456d836adb) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/d1041cb4-38df-46f2-ba03-c8f57474b3eb) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/d1041cb4-38df-46f2-ba03-c8f57474b3eb) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/16dde9c7-4d7f-4e34-aa60-c77b4e755dbc) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/16dde9c7-4d7f-4e34-aa60-c77b4e755dbc) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/5de7a406-5f4a-465c-ba8f-b8bfb26d7a13) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/5de7a406-5f4a-465c-ba8f-b8bfb26d7a13) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/74f12ccc-e400-477a-9b09-34323382bac4) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/74f12ccc-e400-477a-9b09-34323382bac4) completed: **No new issues**

VolCh commented:
I implemented testReflectionPropertyParity() method (with a refactor of testReflectionMethodParity()) to check parity of property. All tests are passed. Then I added FQCN to all parity checks. 

Tests are broken now with 100+ failures!

aik099 commented:
I'm not sure if these are good news or bad news.

Are tests failing due false positives or they really indicate that something not working consistently?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/7ebf107a-1105-4b53-a6b5-6a88f58a14cc) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/7ebf107a-1105-4b53-a6b5-6a88f58a14cc) completed: **No new issues**

VolCh commented:
My last commit emulate original PHP behavior. I explored PHP source about class name related behavior functions & methods: all of them drop leading slash first of all, before parsing or search or save. So I think this is not false positive result.

VolCh commented:
@aik099 what do you think about it?

aik099 commented on src/ReflectionMethod.php:
```diff
@@ -52,7 +52,8 @@ public function __construct(
     ) {
         //for some reason, ReflectionMethod->getNamespaceName in php always returns '', so we shouldn't use it too
-        $this->className        = $className;
+        $this->className        = is_object($className) ? get_class($className) : ltrim($className, '\\');
```
Does this class even support passing an object in `$className` parameter? I'm suspecting it doesn't, because parameter name suggests that. Yeah, I know that PHP built-in reflection supports that.

If it doesn't, then do `ltrim` all the time.

VolCh replied:
\Go\ParserReflection\ReflectionClass supports (documented like string|object) and is_object check present. I decide 'casting' object to string in \Go\ParserReflection\ReflectionMethod & \Go\ParserReflection\ReflectionProperty make them more consist with built-in & \Go\ParserReflection\ReflectionClass and provide expected behavior. Do you agree?

aik099 replied:
I understand now. Since library only supports class names and not objects you've created workaround code to convert object (if somebody gives it by mistake) into a class so that library won't have any issues.

Nice catch, but this (and associated place in `ReflectionProperty` and other places) needs to be moved into separate PR with corresponding tests. I'd better not to mix this with topic of this PR.

aik099 commented on src/ReflectionProperty.php:
```diff
@@ -60,7 +60,8 @@ public function __construct(
         PropertyProperty $propertyNode = null
     ) {
-        $this->className    = $className;
+        $this->className    = is_object($className) ? get_class($className) : ltrim($className, '\\');
```
Does this class even support passing an object in `$className` parameter? I'm suspecting it doesn't, because parameter name suggests that. Yeah, I know that PHP built-in reflection supports that.

If it doesn't, then do `ltrim` all the time.

aik099 commented on tests/ReflectionClassTest.php:
```diff
@@ -47,21 +48,21 @@ public function testGetModifiers($fileName)
 
-        $expectedValue = $refClass->$getterName();
-        $actualValue   = $parsedClass->$getterName();
+        $expectedValue = $refClass->{$getterName}();
```
Is this and next change really needed? 

Should work fine on both PHP5 and PHP7 already.

aik099 commented on tests/ReflectionClassTest.php:
```diff
@@ -47,21 +48,21 @@ public function testGetModifiers($fileName)
+     * @param string $getterName Name of the reflection method to test
      */
     public function testReflectionMethodParity(
+        \ReflectionClass $refClass,
```
I guess `methodCaseProvider` now returns more data, but this particular test doesn't use that extra data?

aik099 commented on tests/ReflectionClassTest.php:
```diff
@@ -290,4 +300,58 @@ protected function getGettersToCheck()
+     * @param string[] $membersToCheck
+     * @return array
+     */
+    protected function caseProvider($membersToCheck)
```
I see you've moved method down, but because of that it's hard to see what you've changed in the method itself.

aik099 commented on tests/ReflectionClassTest.php:
```diff
@@ -290,4 +300,58 @@ protected function getGettersToCheck()
+                            $qcn => [$refClass, new \ReflectionClass($qcn)],
+                            $fqcn => [new ReflectionClass($fqcn), new \ReflectionClass($fqcn)],
+                        ];
+                        foreach ($classes as $caseName => list($parsedClass, $originalClass)) {
```
1. Is this `list(` function usage compatible with all PHP versions where library itself can be used?
2. Please rename `$originalClass` into `$builtinClass`. Maybe it makes sense to rename corresponding parameter in method, where this data provider is used (e.g. `$refMethod` to `$builtinMethod` and such).

VolCh replied:
1. I used list() in PHP3 era :) Changes in 7.0 does not affect this simple use case.
2. ok

aik099 replied:
1 - you could use `list(`, but support for using `list` right within `foreach` construct was added later

loren-osborn replied:
I just checked this, and it appears to work in [PHP 5.5.0 and above](https://3v4l.org/37Ok9). (I don’t think we support anything older than 5.6.0, so I suspect this will be ok. @aik099, can you confirm this?)

aik099 replied:
What I've meant is for @VolCh to find http://php.net/manual/en/control-structures.foreach.php#control-structures.foreach.list, which clearly says that it works since PHP 5.5. That doesn't quite qualify for `I used list() in PHP3 era` :)

Yeah, @loren-osborn , it works according to docs. I haven't used such syntax myself though.

aik099 commented on tests/ReflectionClassTest.php:
```diff
@@ -290,4 +300,58 @@ protected function getGettersToCheck()
+                include_once $fileName;
+                foreach ($reflectionFile->getFileNamespaces() as $fileNamespace) {
+                    foreach ($fileNamespace->getClasses() as $refClass) {
+                        $qcn = ltrim($refClass->getName(), '\\'); // workaround for #80
```
This PR should be ltrimming class name. Why we need to also ltrim in tests?

VolCh replied:
Tests should check that no difference with and without leading backslash between native Reflection* and Go\ParserReflection\Reflection*. It is common consists checking regardless fix in the PR.

aik099 replied:
I'm still not getting it. If this PR normalizes stuff so that there won't be any leading `\` in class names around the library and built-in reflection already does it, then why we need to ltrim it in tests?

loren-osborn commented:
Not that it particularly matters, but the changes in #82 to `tests/` is going to be pretty frustrating to merge with this, as i’ve broken out many of the loops into separate tests. Maybe you want to try to parallel some of that to ease the merge? 😕

...or not.

aik099 commented:
Maybe you can do tests refactoring in separate PR, that can be quickly merged? By the way both PRs are made by you so you're blaming yourself here 😉

loren-osborn commented:
This isn’t my PR. I don’t know what you mean about blaming myself here.

I **HAD** started a test suite refactor branch [here: loren-osborn/parser-reflection:losborn_testDirReorg_orig](https://github.com/loren-osborn/parser-reflection/commits/losborn_testDirReorg_orig) but needed to focus my attention in other direction at the time for various reasons. It was a pretty big change and will be a very hairy merge at this point, but I have borrowed code from it in the last couple PRs I’ve submitted.

I know i’ve had trouble making any practical headway on #82 in my meager 1-2 hours a week on my XP machine that barely runs PHP 5.6, so I’m concerned that breaking off another PR of the complexity of [losborn_testDirReorg_orig](https://github.com/loren-osborn/parser-reflection/tree/losborn_testDirReorg_orig) may seriously derail #82.

aik099 commented:
Now I see, you've (@loren-osborn) mentioned on this PR (authored by @VolCh) that there might be issues for merge.

loren-osborn commented:
@aik099, I did a mental recap of the branch I mentioned above, and realized I had been trying to do too much in one PR, and got stuck on what **should** have been the third of four distinct PRs.

What I had intended for this branch:

1. Reorganize the test directory, separating test classes from stub and support files.
    * The main reason for this was to put support code in a separate area, and make a logical place in the test class directory to test the test-support code.
    * This was in anticipation of refactoring the “meat” of most of the `@dataProvider`s of comparators used for asserts into support classes, keeping the complexity **out** of the test classes.
2. Move the test code into appropriate namespaces.
3. Get to 100% code coverage.
    * This is a big job by itself.
    * At least one of the classes I wanted to test was best done by mocking built-in functions, so I had started work on a bare bones, ApectMock like, tool to do this in a limited capacity.
4. Refactor the **existing** test code. 

I think 1 & 2 naturally go together.  Given the branch is hopelessly out of date, I think the best strategy is to merge it into #82, separate out 1 & 2 together, and 4 by itself into new PRs, pull out 3 to live on it’s own, longer term, branch, and leave the rest behind in #82. This will allow me to leverage the test suite changes I already made in #82.

aik099 commented:
Do whatever you see fit. From PR reviewing point the smaller and on topic PR is the better to me.

> Move the test code into appropriate namespaces.

Aren't test case classes already namespaced?

loren-osborn commented:
@aik099, they share a namespace with the production code. In my opinion this isn’t really appropriate.

aik099 commented:
> ... they share a namespace with the production code. In my opinion this isn’t really appropriate, especially because they don’t live in the production code directories. This also prevents the composer auto loader from working with these files.

There are 2 techniques here:

* like in Symfony and here, where there is `Tests` sub-namespaces within application namespace (e.g. `Vendor\AppName\ClassName` + `Vendor\AppName\Tests\ClassName`)
* like you say (and I like to do) to have `Tests` sub-namespace above application namespace (e.g. `Vendor\AppName\ClassName` + `Tests\Vendor\AppName\ClassName`)

Plus for having `Tests\` in front of namespaces is that you can move auto-loading into `autoload-dev` in `composer.json` and then optimized auto-loader in production environment won't have code for auto-completing test classes.

loren-osborn commented:
@aik099, we’re talking about stuff I did a few months ago, and haven’t really touched since. If I recall currently, I believe I used the first strategy you mentioned above. I also think I was mistaken about composer not working with a namespace spread across multiple directories.

loren-osborn commented:
@aik099, I definitely still have a great deal of refactoring and untangling to do before issuing a PR for this, but I wanted to show you that I have gotten these changes merged into the current code base, and have all tests passing again. Here is my current working branch for this: https://github.com/loren-osborn/parser-reflection/tree/losborn_testDirReorgAndPassThrough

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/d2b82929-eaad-4afa-b219-226f0e0927ba) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/d2b82929-eaad-4afa-b219-226f0e0927ba) completed: **No new issues**

VolCh commented:
@aik099 sorry for the silence. I have been removed object casting to string in ReflectionMethod & Property.

But I don't know what to do with tests. Are so big changes acceptable now? I can make just three small tests to check backslash presence/absence behavior.

aik099 commented:
At this point I'm really thinking, that purpose of the fix (considering amount of changes made) isn't worth it. People were using library with this problem before and haven't complained that much :(

Considering speed of this PR development I need to do full review on each push and that takes a lot of time.

lisachenko commented:
It’s pity, but I have a feeling that we should abandon this project and switch to BetterReflection, because it’s well maintained, whereas we can’t even merge this old PR with huge amount of changes :( Probably, I should work hard on this, but I’ lack of time....

Guys, what do you think? Should we keep working on this library?

aik099 commented:
I'd keep it in current state with policy for accepting small features/bug fixes. It does it job well currently and have wider PHP version support then https://github.com/Roave/BetterReflection.

There are workarounds for known issues, but fixing these issues in the library itself would be too time consuming.

In fact in some reflection-related applications it's better to create thin wrapper around https://github.com/nikic/PHP-Parser instead of using full-featured reflection provided by this or better-reflection library.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/172c7995-9747-45bb-abcb-a0598f875d7d) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/goaop/parser-reflection/inspections/172c7995-9747-45bb-abcb-a0598f875d7d) completed: **No new issues**

VolCh commented:
I really did not think that my PRs would raise the question about abandoning the library. :(

In this case there is no point in globally changing the tests. I have done a minor changes in the locators and class related Reflection* constructors to fix problems with or without leading backslash in any cases. 

If someone use locators directly this PR shouldn't break the code.

I can remove some changes if they are not acceptable by the policy for accepting small features/bug fixes. Also I save previous changes in https://github.com/VolCh/parser-reflection/tree/rewrite-tests

loren-osborn commented:
@aik099, @lisachenko, I've continued devoting what scarce resources I've had to PRs for this project, and am optimistic we may have, at least something to offer merging with BetterReflection, (for instance, we still support PHP 5.5 and 5.6). I hate to admit the time I've invested in recent months may have been for naught, but it's silly "throwing good money after bad".

Has anyone investigated what we have to offer that BetterReflection doesn't do? I'd be eager to hear what you find.

If it's relevant, I've continued to make steady, but glacial progress on PRs for this project.

aik099 commented:
Support for PHP 5.x version at least.

aik099 commented:
@VolCh , is this PR ready for merging?

VolCh commented:
@aik099 Yes, it is ready.

aik099 commented:
Merging, thanks @VolCh .

00eadeyemi commented:
@aik099 do you have any idea when the next release is going to be? I am not sure setting my dependency version to dev-master is a good idea but yet I really need this fix.

aik099 commented:
No idea.

lisachenko commented:
@00eadeyemi patch version 1.4.1

00eadeyemi commented:
@lisachenko Thank you !!!
